### PR TITLE
cancel all long-running tasks

### PR DESCRIPTION
**Pull request by jackstar12**
Not all long-running tasks are cancelled upon shutdown, which caused some issues with `WALLET.cleanup`.

I introduced a few new functions:

- `create_task` basically a wrapper around the asyncio equivalent, adds the task to a list
- `cancel_all_tasks` cancels all tasks previously created with `create_task` (if any are already stopped, `cancel` will do nothing)
- `create_permanent_task` wrapper around `create_task` with `catch_everything_and_restart`

Now all tasks are cancelled on shutdown and `WALLET.cleanup` goes without errors.

### CODE REVIEW

**Review by dni — APPROVED** [review]

**Review by dni — APPROVED** [review]
LGTM

File: lnbits/tasks.py [hunk 1]
```diff
@@ -19,6 +19,23 @@
...
+    return task
+
+
+def create_permanent_task(func):
+    return create_task(catch_everything_and_restart(func))
+
+
+def cancel_all_tasks():
+    for task in tasks:
+        task.cancel()
```
Comment (motorina0) [inline lnbits/tasks.py:37]: - loop body should be wrapped in `try-catch`

File: lnbits/app.py [hunk 2]
```diff
@@ -396,21 +398,13 @@ async def webhook_listener():
 
     @app.on_event("startup")
     async def listeners():
```
Comment (motorina0) [inline lnbits/app.py:400]: - any reason for `async def listeners()` to be `async` now?

File: lnbits/core/tasks.py [hunk 3]
```diff
@@ -20,24 +25,13 @@
...
-async def register_killswitch():
+def register_killswitch():
     """
     Registers a killswitch which will check lnbits-status repository
     for a signal from LNbits and will switch to VoidWallet if the killswitch is triggered.
     """
     logger.debug("Starting killswitch task")
-    global killswitch
-    killswitch = asyncio.create_task(killswitch_task())
-
```
Comment (motorina0) [inline lnbits/core/tasks.py:31]: - any reason for `killswitch: Optional[asyncio.Task] = None` to still be present?
- see: https://github.com/lnbits/lnbits/blob/2928209659950f27cadd4701f5a10c84a75efe2b/lnbits/core/tasks.py#L24

File: lnbits/core/tasks.py [hunk 4]
```diff
@@ -73,16 +67,6 @@ async def register_watchdog():
     # watchdog = asyncio.create_task(watchdog_task())
 
 
-async def unregister_watchdog():
```
Comment (motorina0) [inline lnbits/core/tasks.py:76]: - since `watchdog` was also converted please adapt the commented code to the new paradigm

https://github.com/lnbits/lnbits/blob/2928209659950f27cadd4701f5a10c84a75efe2b/lnbits/core/tasks.py#L67

File: lnbits/server.py [hunk 5]
```diff
@@ -1,14 +1,10 @@
-import uvloop
-from uvicorn.supervisors import ChangeReload
-
-uvloop.install()
```
Comment (motorina0) [inline lnbits/server.py:4]: - please help me understand, what was this line doing and why it is not required now

File: lnbits/tasks.py [hunk 6]
```diff
@@ -19,6 +19,23 @@
 
 from .core import db
 
+tasks: List[asyncio.Task] = []
+
+
+def create_task(coro_or_fut):
```
Comment (motorina0) [inline lnbits/tasks.py:25]: - name nitpick: change `coro_or_fut` to `coro_or_func` (consistent with the below use and more clear)

File: lnbits/wallets/cln.py [hunk 7]
```diff
@@ -23,23 +22,9 @@
...
-    return run
-
-
-def _pay_invoice(ln, payload):
-    return ln.call("pay", payload)
-
-
-def _paid_invoices_stream(ln, last_pay_index):
-    return ln.waitanyinvoice(last_pay_index)
+async def run_sync(func) -> Any:
```
Comment (motorina0) [inline lnbits/wallets/cln.py:25]: - how is this related to `asyncio.create_task` improvements?

Comment (dni) [reply lnbits/wallets/cln.py:25]: https://github.com/lnbits/lnbits/pull/1858

File: lnbits/tasks.py (hunk 1 above)
Comment (jackstar12) [reply lnbits/tasks.py]: `task.cancel` doesnt throw, even if its already canceled

File: lnbits/tasks.py (hunk 6 above)
Comment (jackstar12) [reply lnbits/tasks.py]: `coro_or_fut` is the same as asyncio uses themselves. It doesnt have to be the same as below because it isn't, because the below takes a func which creates a coroutine

File: lnbits/wallets/cln.py (hunk 7 above)
Comment (jackstar12) [reply lnbits/wallets/cln.py]: this fix was required to get tests to pass. cln would otherwise get stuck

File: lnbits/server.py (hunk 5 above)
Comment (jackstar12) [reply lnbits/server.py]: see commit description: https://github.com/lnbits/lnbits/pull/1860/commits/ec2aed001d79bec044fc46644bc55212f18c0a48

**Comment by jackstar12** [comment]
another interesting side effect of this is that extensions are loaded inside tests aswell, but I dont think this should be considered an issue

File: lnbits/tasks.py (hunk 6 above)
Comment (jackstar12) [reply lnbits/tasks.py]: `coro` would be the most accurate in this case

**Review by motorina0 — APPROVED** [review]
LGTM

great job @jackstar12 !

**Comment by dni** [comment]
> > another interesting side effect of this is that extensions are loaded inside tests aswell, but I dont think this should be considered an issue
> 
>     * only extensions listed in `LNBITS_EXTENSIONS_DEFAULT_INSTALL` should be installed by default
> 
>     * if this value is empty no extension will be installed

good catch, dont install default exts for test when run locally! :)
https://github.com/lnbits/lnbits/pull/1861
